cloud-aws/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/AwsSubnetIgwExplorerTest.java

```diff
@@ -83,54 +89,201 @@ public void testWithIgwButNoAssociation() {
 
         assertFalse(hasInternetGateway);
     }
 
     @Test
     public void testWithNoAssociationAndNoIgw() {
         DescribeRouteTablesResult describeRouteTablesResult = new DescribeRouteTablesResult();
-        RouteTable routeTable = new RouteTable();
+        RouteTable routeTable = new RouteTable().withVpcId(VPC_ID);
         Route route = new Route();
         routeTable.setRoutes(List.of(route));
         RouteTableAssociation routeTableAssociation = new RouteTableAssociation();
         routeTableAssociation.setSubnetId(DIFFERENT_SUBNET_ID);
         routeTable.setAssociations(List.of(routeTableAssociation));
         describeRouteTablesResult.setRouteTables(List.of(routeTable));
 
-        boolean hasInternetGateway = awsSubnetIgwExplorer.hasInternetGatewayOfSubnet(describeRouteTablesResult, SUBNET_ID);
+        boolean hasInternetGateway = awsSubnetIgwExplorer.hasInternetGatewayOfSubnet(describeRouteTablesResult, SUBNET_ID, VPC_ID);
 
         assertFalse(hasInternetGateway);
     }
 
     @Test
     public void testWithRouteButNoAssociations() {
         DescribeRouteTablesResult describeRouteTablesResult = new DescribeRouteTablesResult();
-        RouteTable routeTable = new RouteTable();
+        RouteTable routeTable = new RouteTable().withVpcId(VPC_ID);
         Route route = new Route();
         routeTable.setRoutes(List.of(route));
         describeRouteTablesResult.setRouteTables(List.of(routeTable));
 
-        boolean hasInternetGateway = awsSubnetIgwExplorer.hasInternetGatewayOfSubnet(describeRouteTablesResult, SUBNET_ID);
+        boolean hasInternetGateway = awsSubnetIgwExplorer.hasInternetGatewayOfSubnet(describeRouteTablesResult, SUBNET_ID, VPC_ID);
+
+        assertFalse(hasInternetGateway);
+    }
+
+    @Test
+    public void testWithMainRoutTableAssociationAndNoIgw() {
+        DescribeRouteTablesResult describeRouteTablesResult = new DescribeRouteTablesResult();
+        RouteTable routeTable = new RouteTable().withVpcId(VPC_ID);
+        Route route = new Route()
+                .withDestinationCidrBlock(INTERNAL_DESTINATION_CIDR_BLOCK);
+        routeTable.setRoutes(List.of(route));
+        RouteTableAssociation routeTableAssociation = new RouteTableAssociation()
+                .withMain(Boolean.TRUE);
+        routeTable.setAssociations(List.of(routeTableAssociation));
+        describeRouteTablesResult.setRouteTables(List.of(routeTable));
+
+        boolean hasInternetGateway = awsSubnetIgwExplorer.hasInternetGatewayOfSubnet(describeRouteTablesResult, SUBNET_ID, VPC_ID);
 
         assertFalse(hasInternetGateway);
     }
 
+    @Test
+    public void testWithMainRoutTableAssociationAndIgw() {
+        DescribeRouteTablesResult describeRouteTablesResult = new DescribeRouteTablesResult();
+        RouteTable routeTable = new RouteTable().withVpcId(VPC_ID);
+        Route route = new Route()
+                .withGatewayId(GATEWAY_ID)
```

Comment:
missing case (could be invalid though):
what if the gateway is there but is not pointing to an open CIDR block

Reply:
I think that's covered by this test case: https://github.com/hortonworks/cloudbreak/blob/7d83096e17dd848ee9a3148643830f86660d2253/cloud-aws/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/AwsSubnetIgwExplorerTest.java#L66